JWT: Fix organization switching for JWT auth plugin

QuentinBisson opened the pull request:
**What is this feature?**

This fixes the recently introduced org mapping for the jwt.auth that does not support organization switching.

This fix makes sure we use the org id extracted by the authn.Service is used instead

**Why do we need this feature?**

Org mapping was recently introduced (by me) in the JWT plugin but the resulting identity is configured without the request orgId parameter. This causes an infinite redirection loop (due to the code in the org redirect middleware) because the context org Id is always 1.

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

QuentinBisson committed:
Fix organization switching for jwt auth

QuentinBisson committed:
add-support-for-org-id-header

QuentinBisson commented on pkg/services/authn/authnimpl/service.go:
```diff
@@ -537,14 +537,29 @@ func parseNamespace(path string) string {
 
-// name of query string used to target specific org for request
+// name of the query string used to target specific org for request
 const orgIDTargetQuery = "targetOrgId"
```
Hey @mgyongyosi I was testing the original PR in my environment and I stumbled upon this targetOrgId query param and this PR https://github.com/grafana/grafana/pull/60528/files.

I tried to set the targetOrgId to match the orgId query param and this actually solved my issue.

The PR I linked is talking about some authentication requests and I think the JWT requests might have been in them :D 

Anyway, I have done some shallow research and I only find a reference to the ?targetOrgId query param in one typescript file so I'm wondering it that is not some legacy that should be renamed to orgId instead of targetOrgId everywhere?

What do you think about this?

QuentinBisson commented:
This is still needed

QuentinBisson commented:
No worry @cinaglia we make this somehow usable via a side-car proxy for now but i'll admit having a propre fix would bé awesome :D

QuentinBisson commented:
Yes this is what I wanted to fix but I did not test this locally but in my environnement which incluses reverse proxies so it could explain why it worked ony end. I'll try to run this locally this week. Thanks for looking into this :)

QuentinBisson commented:
I tried to replicate this locally but I did not manage to 🤔 

So checking this again now that i had time, yes i wanted to fix the orgId uery string redirection when coming from outside Grafana so I can link to dashboard in other organizations from our alerting system. 

The organization switcher does work on my machine with the use of the org mapping which I did not see in your configuration. Maybe that explain why it does not work?

cinaglia commented:
@QuentinBisson Just wanted to follow up on this. Were you still hoping to contribute a fix for this? Thanks!

QuentinBisson committed:
fix(auth): Fix JWT authentication organization switching

This commit addresses the JWT authentication organization switching bug that
caused infinite redirect loops when users tried to access Grafana with a
specific organization via URL parameters (e.g., ?orgId=2).

## Root Cause
The JWT authentication client was inconsistent with other social auth clients
(OAuth, SAML) in how it handled organization assignment:

1. **Missing OrgID Assignment**: The JWT client wasn't setting the requested
   OrgID from the authn.Request, causing the org redirect middleware to fail
   when users tried to switch organizations.

2. **Duplicate Organization Logic**: The JWT client had custom organization
   assignment logic instead of leveraging the existing sync hooks that other
   auth clients use for consistent organization management.

## Solution
1. **Simplified JWT Client**: Removed custom organization assignment logic
   and now rely on the existing  which:
   - Handles organization role synchronization
   - Validates user access to requested organizations
   - Falls back to valid organizations when needed
   - Updates the user's default org appropriately

2. **Consistent Behavior**: JWT authentication now follows the same pattern
   as OAuth and other social auth providers:
   - Set the requested OrgID in the identity
   - Let sync hooks handle validation and assignment
   - Remove code duplication between auth clients

3. **Enhanced Query Parameter Support**: Added support for both 'orgId' and
   'targetOrgId' query parameters for backward compatibility.

## Key Changes
- : Simplified org assignment logic
- : Added orgId query param support
- : Updated test expectations

This fix ensures JWT users can now switch organizations via URL parameters
without infinite redirects, matching the behavior of other auth providers.

QuentinBisson committed:
test(auth): Add comprehensive JWT organization switching tests

This commit adds comprehensive tests to verify that the JWT authentication
organization switching fix works correctly. The tests cover:

1. **JWT Organization Parameter Handling:**
   - Verifies ?orgId=2 correctly sets identity OrgID to 2
   - Verifies ?targetOrgId=3 correctly sets identity OrgID to 3
   - Tests fallback behavior when no org parameter is provided
   - Tests behavior with and without org mapping configuration

2. **Query Parameter Parsing:**
   - Verifies both 'orgId' and 'targetOrgId' parameters are parsed correctly
   - Confirms 'orgId' takes precedence over 'targetOrgId'
   - Tests edge cases with no parameters or invalid values

These tests ensure that JWT users can now switch organizations via URL
parameters without infinite redirects, and that the fix maintains
consistency with other authentication providers.

Test results show all scenarios pass, confirming the organization
switching bug has been properly resolved.

QuentinBisson commented:
@cinaglia I think this should fix the issue by relying on the sync hook but let me know if that;s not enough

QuentinBisson committed:
fix: JWT authentication organization switching

- Fix JWT client to properly set identity.OrgID from request parameters
- Prevent infinite redirect loops when switching orgs with ?orgId parameter
- Use configuration-based sync control via SkipOrgRoleSync flag
- Handle edge cases where JWT has no org claims but URL-based switching needed
- Let sync hooks handle organization validation and assignment logic
- Add comprehensive tests for organization switching scenarios

Fixes issue where JWT auth users couldn't switch organizations via URL
parameters, causing infinite redirect loops in org redirect middleware.

QuentinBisson commented:
Not stale

QuentinBisson commented:
I updated the PR, it is still waiting for review

QuentinBisson commented:
Hey this is not stale :)
